### DISCUSSION TITLE
Tracing/drawing a protein (chain by chain)

### DISCUSSION THREAD

**Discussion by heejongkim** [BradyAJohnston/MolecularNodes#192] [Ideas]
Hi Brady,

Thank you for releasing the new version with the capability of importing EM density. I'm so excited to play around and hope to see slicing and other functionalities to come.

By the way, I have a question about animating a protein by tracing/drawing ribbon from n-term to c-term.
I remember i watched someone did it before but I can't reference that again and have no clue how to accomplish that.

If you can share any hints, that would be fantastic.

Thank you.

best,
heejong

**Comment by BradyAJohnston** [comment]
You can use the curve generated by the `MOL_style_ribbon_protein` node to animate the 'drawing' of the backbone. This setup should be enough to get you started.

![image](https://user-images.githubusercontent.com/36021261/230839311-a2285ea6-1f24-4ead-b065-7f12f9f212e2.png)

**Reply by heejongkim** [reply]
Thanks for the setup. Fantastic!

I was testing the setup coupling with chain selection and noticed that it's drawing not really linearly but it starts with any starting residues in parallel if model has discontinuity. Do i need to put more restriction on MOL_animate_value to accomplish true linear tracing?

Thank you!

**Reply by BradyAJohnston** [reply]
Do you mean that the structure is missing some residues and so it isn't a continuous curve? Do you have a PDB example that I can test on?

**Reply by heejongkim** [reply]
Any PDB is fine.
Let say you have a single protein with 1000 residues. You delete 500-600 in pdb thru chimera, import pdb, and do the workflow above.
You will see the tracing happening from 1 to 499 and 601 to 1000 in parallel.

**Reply by BradyAJohnston** [reply]
Ah I understand what you mean. Here is a way you can do it via using a `Map Range` node and utilising the `res_id` attribute. As you change the `Value` slider, it will trim off residues above that number.

![image](https://user-images.githubusercontent.com/36021261/232181838-cbad6387-7f5f-4915-bf44-c01f79d83ea0.png)

You can combine it with a `Attribute Statistic` node to instead create a `0 -> 1` slider that will affect proteins of varying lengths equally.

![image](https://user-images.githubusercontent.com/36021261/232181867-b7b4d314-bbf1-4b56-bdf6-4ffd14d098db.png)

This might be something that I can build into the ribbon node itself, seems like it could be useful for some animations.

**Reply by heejongkim** [reply]
Awesome. I think it would be awesome addition to your existing nodes.

I'm a little bit lost where I can get res_id and x + 1. I can't really find them from "Add".
Can you please give me a hit for that?

Also, from the workflow above, if i want to set the start and end in timeline with keyframe, which attribute or node should i utilize?

Thank you!

**Reply by BradyAJohnston** [reply]
Ah, the `x + 1` is just a math node that I have collapsed and renamed for convenience. Sorry for the confusion!

![image](https://user-images.githubusercontent.com/36021261/232956853-20518e23-b5b1-4ef3-9dab-5b9a40cef155.png)

Above is an example using the `MOL_animate_value` node. That allows you to set the start and end frame, which will be when the animation starts and ends.

**Reply by heejongkim** [reply]
Ooops. I feel dumb now. haha.
Thanks for the clarification and screenshot.

Let me play around a little bit and get back to you.
From this workflow, ribbon protein's selection input is already occupied by Map Range so I wonder how I can choose a specific chain by MOL_sel_chain.

Thank you!

p.s. I'm so excited to see that secondary structure depicting capability! That's so great!

**Reply by BradyAJohnston** [reply]
You can combine the `Map Range` with the `MOL_sel_chain` using a `Boolean Math` node set to `And`. The result of the `Map Range` is just `0 -> 1` which is the same as a boolean `False -> True` and so the two selections can be combined.

**Reply by BradyAJohnston** [reply]
And yes I'm excited about the secondary structure as well!!

**Reply by heejongkim** [reply]
awesome! Thank you so much!!
Hoping I can construct this to animate first and swap the style with secondary structure to make it appealing to structural people.

Have a good day!

**Reply by heejongkim** [reply]
And I can give infinite number of votes for this tracing node!
